cuML/src/tsne/TSNE.h

```diff
@@ -0,0 +1,121 @@
+
+using namespace ML;
+
+#include "cuda_utils.h"
+#include <cuda_runtime.h>
+#include <math.h>
+#define cuda_free(x)        CUDA_CHECK(cudaFree(x))
+#define cuda_malloc(x, n)   MLCommon::allocate(x, n);
+
+
+template <typename Type>
+void cuda_calloc(Type *&ptr, size_t n, Type val) {
+    // From ml-prims / src / utils.h
+    // Just allows easier memsetting
+    CUDA_CHECK(cudaMalloc((void **)&ptr, sizeof(Type) * n));
+    CUDA_CHECK(cudaMemset(ptr, val, sizeof(Type) * n));
+}
+
+
+__global__ void upper_lower_assign( float * __restrict__ sigmas,
+                                    float * __restrict__ lower_bound,
+                                    float * __restrict__ upper_bound,
+                                    const float * __restrict__ perplexity,
+                                    const float target_perplexity,
+                                    const int n)
+{
+    const int i = threadIdx.x + blockIdx.x * blockDim.x;
+    if (i < n) {
+        if (perplexity[i] > target_perplexity)
+            upper_bound[i] = sigmas[i];
+        else
+            lower_bound[i] = sigmas[i];
+        sigmas[i] = (upper_bound[i] + lower_bound[i]) * 0.5f;
+    }
+}
+
+
+#include "distance/euclidean.h"
+#include "linalg/power.h"
+#include "linalg/matrix_vector_op.h"
+template<typename Type>
+void TSNE::compute_Pij( float * __restrict__ Pij;
+                        const Type * __restrict__ X,
+                        float * __restrict__ sigmas;
+                        const int n,
+                        const int p,
+                        cudaStream_t * stream,
+                        const int full_size,
+                        float * __restrict__ sigma_squared)
+{   
+    // Workspace needed to get rowNorm(X)
+    size_t lwork = n * sizeof(Type);
+    float *work; cuda_malloc(work, n);
+
+    // From distances / euclidean algo1
+                        // row  col col  in  in  out  sqrt   work  lwork  norm  stream
+    Distance::euclideanAlgo1(n,  p,  p,  X,  X,  Pij, false, work, lwork,   2,  stream);
```

Comment:
Call the exposed distance API function rather than the internal algorithm implementation:

https://github.com/rapidsai/cuml/blob/branch-0.7/ml-prims/test/distance_base.h#L134

Reply:
Yep good point! Will fix that